File: articles/mathematical-modeling-in-ecology-and-sustainability/c/sustainability_scenario_summary.c
#include <stdio.h>
#include <stdlib.h>

typedef struct {
    const char *key;
    const char *scenario_name;
    double initial_stock;
    double growth_rate;
    double carrying_capacity;
    double extraction;
    double climate_stress;
    int years;
    double minimum_stock;
} ResourceScenario;

typedef struct {
    double final_stock;
    double minimum_observed_stock;
    double minimum_resilience_margin;
    int threshold_breach;
} Evaluation;

Evaluation evaluate(ResourceScenario scenario) {
    double stock = scenario.initial_stock;
    double effective_growth = scenario.growth_rate * (1.0 - scenario.climate_stress);
    double min_stock = stock;
    double min_margin = stock - scenario.minimum_stock;

    for (int year = 0; year < scenario.years; ++year) {
        double regeneration = effective_growth * stock * (1.0 - stock / scenario.carrying_capacity);
        stock = stock + regeneration - scenario.extraction;
        if (stock < 0.0) {
            stock = 0.0;
        }
        if (stock < min_stock) {
            min_stock = stock;
        }
        double margin = stock - scenario.minimum_stock;
        if (margin < min_margin) {
            min_margin = margin;
        }
    }

    Evaluation result = {stock, min_stock, min_margin, min_stock < scenario.minimum_stock};
    return result;
}

int main(void) {
    ResourceScenario scenarios[] = {
        {"baseline", "Baseline managed use", 420.0, 0.24, 800.0, 36.0, 0.04, 25, 250.0},
        {"high_extraction", "High extraction pressure", 420.0, 0.24, 800.0, 64.0, 0.04, 25, 250.0},
        {"climate_stress", "Climate stress with lower regeneration", 420.0, 0.24, 800.0, 42.0, 0.22, 25, 250.0},
        {"restoration_pathway", "Restoration and reduced extraction", 420.0, 0.28, 860.0, 24.0, 0.03, 25, 250.0},
        {"adaptive_management", "Adaptive use with monitoring trigger", 420.0, 0.25, 820.0, 32.0, 0.08, 25, 250.0}
    };

    printf("key,final_stock,minimum_observed_stock,minimum_resilience_margin,threshold_breach\n");

    for (int i = 0; i < 5; ++i) {
        Evaluation eval = evaluate(scenarios[i]);
        printf("%s,%.6f,%.6f,%.6f,%s\n",
               scenarios[i].key,
               eval.final_stock,
               eval.minimum_observed_stock,
               eval.minimum_resilience_margin,
               eval.threshold_breach ? "true" : "false");
    }

    fprintf(stderr, "sustainability_scenario_summary complete\n");
    return EXIT_SUCCESS;
}

File: articles/mathematical-modeling-in-ecology-and-sustainability/fortran/sustainability_scenario_summary.f90
program sustainability_scenario_summary
  implicit none

  integer, parameter :: n = 5
  character(len=32) :: keys(n)
  real(8) :: initial_stock(n), growth_rate(n), capacity(n), extraction(n)
  real(8) :: climate_stress(n), minimum_stock(n)
  real(8) :: final_stock(n), min_stock(n), min_margin(n)
  integer :: years(n), i
  logical :: breach

  keys = (/ 'baseline                      ', 'high_extraction               ', &
            'climate_stress                ', 'restoration_pathway           ', &
            'adaptive_management           ' /)
  initial_stock = (/ 420.0d0, 420.0d0, 420.0d0, 420.0d0, 420.0d0 /)
  growth_rate = (/ 0.24d0, 0.24d0, 0.24d0, 0.28d0, 0.25d0 /)
  capacity = (/ 800.0d0, 800.0d0, 800.0d0, 860.0d0, 820.0d0 /)
  extraction = (/ 36.0d0, 64.0d0, 42.0d0, 24.0d0, 32.0d0 /)
  climate_stress = (/ 0.04d0, 0.04d0, 0.22d0, 0.03d0, 0.08d0 /)
  years = (/ 25, 25, 25, 25, 25 /)
  minimum_stock = (/ 250.0d0, 250.0d0, 250.0d0, 250.0d0, 250.0d0 /)

  print '(A)', 'key final_stock minimum_observed_stock minimum_resilience_margin threshold_breach'

  do i = 1, n
    call evaluate_resource(initial_stock(i), growth_rate(i), capacity(i), extraction(i), &
      climate_stress(i), years(i), minimum_stock(i), final_stock(i), min_stock(i), min_margin(i))
    breach = min_stock(i) < minimum_stock(i)
    print '(A,1X,F10.3,1X,F10.3,1X,F10.3,1X,L1)', trim(keys(i)), &
      final_stock(i), min_stock(i), min_margin(i), breach
  end do

contains

  subroutine evaluate_resource(initial, growth, carrying_capacity, extraction_rate, stress, &
      n_years, threshold_stock, final_value, min_value, min_margin_value)
    implicit none
    real(8), intent(in) :: initial, growth, carrying_capacity, extraction_rate, stress
    integer, intent(in) :: n_years
    real(8), intent(in) :: threshold_stock
    real(8), intent(out) :: final_value, min_value, min_margin_value
    integer :: year
    real(8) :: stock, effective_growth, regeneration

    stock = initial
    effective_growth = growth * (1.0d0 - stress)
    min_value = stock
    min_margin_value = stock - threshold_stock

    do year = 1, n_years
      regeneration = effective_growth * stock * (1.0d0 - stock / carrying_capacity)
      stock = max(0.0d0, stock + regeneration - extraction_rate)
      min_value = min(min_value, stock)
      min_margin_value = min(min_margin_value, stock - threshold_stock)
    end do

    final_value = stock
  end subroutine evaluate_resource

end program sustainability_scenario_summary
